lib/trento.pm

```diff
@@ -63,12 +66,97 @@ use constant TRENTO_AZ_PREFIX => 'openqa-trento';
+    my ($self, $work_dir) = @_;
+
+    enter_cmd "cd ${work_dir}";
+    my $gitlab_token = get_var(TRENTO_GITLAB_TOKEN => get_required_var('_SECRET_TRENTO_GITLAB_TOKEN'));
+
+    # Script from a release
+    if (get_var('TRENTO_DEPLOY_VER')) {
+
+        # Clean up whatever is present in the indicated working folder
+        # of the qcow2 image, to avoid confision
+        script_run('rm -rf *');
+
+        # Get some 'coordinates' from the test settings
+        my $ver = get_var('TRENTO_DEPLOY_VER');
+        my $gitlab_repo = get_var(TRENTO_GITLAB_REPO => 'gitlab.suse.de/qa-css/trento');
+        my @gitlab_url = split('/', $gitlab_repo);
+        my $gitlab_namespace = $gitlab_url[-2];
+        my $gitlab_project = $gitlab_url[-1];
+
+        my $curl_cmd = 'curl -s ' .
+          "--header \"PRIVATE-TOKEN: ${gitlab_token}\" ";
+        my $gitlab_api = 'https://gitlab.suse.de/api/v4/projects';
+
+        # Get the ID of the requested project
+        my $repo_id_cmd = $curl_cmd .
+          " \"${gitlab_api}/${gitlab_namespace}%2F${gitlab_project}\"" .
+          " | jq -r '.id'";
+        my $repo_id = script_output($repo_id_cmd);
+        # hotfix: harcoded here as not able to strip color codes added also if using -r
+        $repo_id = '7182';
+
+        my $repo_url = "${gitlab_api}/${repo_id}";
+        my $rel_api_url = "${repo_url}/releases/v${ver}";
+
+        # Get the file name of the release archive
+        my $ver_artifact_cmd = $curl_cmd .
+          " ${rel_api_url} " .
+          "| jq -r '.assets.sources[]|select(.format == \"tar.gz\")" .
+          "|.url'|rev|cut -d/ -f1|rev";
+        my $ver_artifact = script_output($ver_artifact_cmd);
+
+        # Get the commit ID of the release
+        my $commit_id_cmd = $curl_cmd .
+          " ${rel_api_url} " .
+          "| jq -r .commit.id";
+        my $commit_id = script_output($commit_id_cmd);
+
+        # Download the release archive
+        my $download_api_url = "${repo_url}/repository/archive.tar.gz?sha=${commit_id}";
+        my $download_cmd = $curl_cmd .
+          " \"${download_api_url}\"" .
+          " --output ${ver_artifact} ";
+        assert_script_run($download_cmd);
+
+        # Extract and test file presence
+        my $tar_cmd = "tar xvf ${ver_artifact} --strip-components=1";
+        assert_script_run($tar_cmd);
+        script_run('ls -lai');
+    }
+
+    # script from gitlab
+    else {
+        my $git_branch = get_var(TRENTO_GITLAB_BRANCH => 'master');
+
+        # Test that the token in worker.ini and the one in the qcow2 match
+        my $qcow_token_cmd = 'git config --get remote.origin.url' .
+          '|cut -d@ -f1|cut -d: -f3';
+        my $qcow_token = script_output($qcow_token_cmd);
+        die "Invalid gitlab token" unless $gitlab_token eq $qcow_token;
+
+        # Switch branch and get latest
+        assert_script_run("git checkout " . $git_branch);
+        assert_script_run("git pull origin " . $git_branch);
```

Comment:
Won't this make repeatability of the test complicated? If you run it once and then restart it, there's no guarantee that you're running the same code from the trento repo.

Reply:
yes, and for this reason I created TRENTO_DEPLOY_VER
On the other hand, the 'pull' approach help a lot during the deployment script and test development, and allow to use custom deployment script when debugging failures